File: ios/BiangBiang Hanzi/Services/HistoryStore.swift
import Foundation

/// Pure history-list mutation rules. No persistence, no UI.
enum HistoryStore {
    /// Silent safety cap. Never surfaced in UI.
    static let safetyCap = 500

    /// Prepend `entry` (newest first), skipping if it duplicates the
    /// most-recent entry by original text + variant. Evicts oldest entries
    /// beyond `safetyCap`.
    static func insert(
        _ entry: HistoryEntry,
        into list: [HistoryEntry]
    ) -> [HistoryEntry] {
        if let newest = list.first,
           newest.original == entry.original,
           newest.variant == entry.variant
        {
            return list
        }
        var result = list
        result.insert(entry, at: 0)
        if result.count > safetyCap {
            result.removeLast(result.count - safetyCap)
        }
        return result
    }

    static func delete(
        id: UUID,
        from list: [HistoryEntry]
    ) -> [HistoryEntry] {
        list.filter { $0.id != id }
    }

    static func clear() -> [HistoryEntry] {
        []
    }
}

File: android/app/src/main/java/dev/veeso/biangbianghanzi/services/HistoryStore.kt
package dev.veeso.biangbianghanzi.services

import dev.veeso.biangbianghanzi.models.HistoryEntry

/** Pure history-list mutation rules. No persistence, no UI. */
object HistoryStore {
    /** Silent safety cap. Never surfaced in UI. */
    const val SAFETY_CAP = 500

    fun insert(
        entry: HistoryEntry,
        list: List<HistoryEntry>,
    ): List<HistoryEntry> {
        val newest = list.firstOrNull()
        if (newest != null &&
            newest.original == entry.original &&
            newest.variant == entry.variant
        ) {
            return list
        }
        val result = ArrayList<HistoryEntry>(list.size + 1)
        result.add(entry)
        result.addAll(list)
        return if (result.size > SAFETY_CAP) {
            result.subList(0, SAFETY_CAP).toList()
        } else {
            result
        }
    }

    fun delete(
        id: String,
        list: List<HistoryEntry>,
    ): List<HistoryEntry> = list.filterNot { it.id == id }

    fun clear(): List<HistoryEntry> = emptyList()
}
